### PR TITLE
UNR-3761/Handover-Gym-Automation

### PR DESCRIPTION
**Pull request by AndreiTeodorLazar**
#### Description
This PR aims to automate the Handover Gym.

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/UNR-3761/SpatialTestHandover/SpatialTestHandover.cpp [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+void ASpatialTestHandover::BeginPlay()
+{
+	Super::BeginPlay();
+
+	// Step Definition to move the HandoverCube to the corresponding test location.
+	FSpatialFunctionalTestStepDefinition MoveCubeStepDefinition;
+	MoveCubeStepDefinition.bIsNativeDefinition = true;
+	MoveCubeStepDefinition.NativeStartEvent.BindLambda([this]()
+		{
+			if (LocationIndex >= TestLocations.Num() || LocationIndex < 0)
```
Comment (ImprobableNic) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/UNR-3761/SpatialTestHandover/SpatialTestHandover.cpp:58]: Don't you need to initialize the LocationIndex to 0 at the start of BeginPlay ?

Comment (ImprobableNic) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/UNR-3761/SpatialTestHandover/SpatialTestHandover.cpp:58]: NVM, I see it done below.

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/UNR-3761/SpatialTestHandover/SpatialTestHandover.cpp [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+	Author = "Andrei";
+	Description = TEXT("Test Actor handover");
+
+	SpawnLocation = FVector(-500.0f, -500.0f, 50.0f);
+
+	TestLocations.Add(FVector(500.0f, -500.0f, 50.0f));
+	TestLocations.Add(FVector(500.0f, 500.0f, 50.0f)); 
+	TestLocations.Add(FVector(-500.0f, 500.0f, 50.0f));
+
+	ExpectedAuthoritativeServer.Add(1);
```
Comment (ImprobableNic) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/UNR-3761/SpatialTestHandover/SpatialTestHandover.cpp:42]: This relies on knowing in advance how the LBS will allocate the virtual worker Ids, and may be brittle. You could instead interrogate the NetDriver's LoadBalancingStrategy, like I do in https://github.com/spatialos/UnrealGDK/pull/2447 -> SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/DebugInterface/TestDebugInterface.cpp

This is not ideal either, and assumes you know which LBS is set on the map, but is not tied to the area allocation scheme anymore.

Comment (AndreiTeodorLazar) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/UNR-3761/SpatialTestHandover/SpatialTestHandover.cpp:42]: Yes that is correct, I will take a look and modify this accordingly, thank you for the suggestion!

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/UNR-3761/SpatialTestHandover/SpatialTestHandover.cpp [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+			FinishStep();
+		});
+
+	// Step Definition to check if the correct server is authoritative over the HandoverCube.
+	FSpatialFunctionalTestStepDefinition CheckAuthorityStepDefinition;
+	CheckAuthorityStepDefinition.bIsNativeDefinition = true;
+	CheckAuthorityStepDefinition.TimeLimit = 10.0f;
+	CheckAuthorityStepDefinition.NativeTickEvent.BindLambda([this](float DeltaTime)
+		{
+			// If a Server Worker holds a lock over the HandoverCube, then it should have authority over it, if not, let the Load Balacning decide.
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/UNR-3761/SpatialTestHandover/SpatialTestHandover.cpp:75]:
```suggestion
			// If a Server Worker holds a lock over the HandoverCube, then it should have authority over it, if not, let the Load Balancer decide.
```

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/UNR-3761/SpatialTestHandover/SpatialTestHandover.cpp [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+				{
+					// Reset the LocationIndex to allow multiple executions of the test.
+					LocationIndex = 0;
+					FinishStep();
+				}
+			}
+		}, 5.0f);
+
+	// Check that Server 1 is authoritative over the HandoverCube.
+	AddStepFromDefinition(CheckAuthorityStepDefinition, FWorkerDefinition::AllServers);
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/UNR-3761/SpatialTestHandover/SpatialTestHandover.cpp:130]: One concern I have here is that if the actor move command, doesn't actually move the cube to a new server authoritative area, then no handover will take place, but the test will all still pass.
Can we add some validation that handover is actually occurring? Perhaps tap into OnAuthGained/Lost?

Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/UNR-3761/SpatialTestHandover/SpatialTestHandover.cpp:130]: Or when we validate authority, we validate authority onto an expectant worker, that increments with each step

**Review by m-samiec — COMMENTED** [review]
One comment, looking good

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/UNR-3761/SpatialTestHandover/SpatialTestHandover.cpp (hunk 4 above)
Comment (AndreiTeodorLazar) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/UNR-3761/SpatialTestHandover/SpatialTestHandover.cpp:130]: Oh, yes this is a consequence of querying the load balancer that I did not take into account, will update this, thank you very much for the catch!

**Review by ImprobableNic — APPROVED** [review]
I was expecting you to introspect the LoadBalancingStrategy to determine the link between workerId and spatial location, bit Mike's suggestion does the job.

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/UNR-3761/SpatialTestHandover/SpatialTestHandover.cpp [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+			else
+			{
+				// If there is a locking server, then the HandoverCube should have not changed authority, therefore we early out if it has changed authority.
+				if (HandoverCube->AuthorityChanges != AuthorityChanges)
+				{
+					return;
+				}
+			}
+
+			// Update the number of AuthorityChanges the HandoverCube had so far.
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/UNR-3761/SpatialTestHandover/SpatialTestHandover.cpp:93]:
```suggestion
			// Update the number of expected AuthorityChanges the HandoverCube should have had so far.
```

**Review by m-samiec — APPROVED** [review]

**Pull request merged**